src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1859,7 +1859,7 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
       if (folded ne tree) return adaptConstant(folded, folded.tpe.asInstanceOf[ConstantType])
       // drop type if prototype is Unit
       if (pt isRef defn.UnitClass)
-        return tpd.Block(tree :: Nil, Literal(Constant(())))
+        return adapt(tpd.Block(tree :: Nil, Literal(Constant(()))), pt)
```

Comment:
This is not enough, the Block itself is not inlinable, so adapting the block doesn't help:
```scala
object X {
  inline def foo: Int = 1

  def test: Unit = {
    X.foo
  }
}
```
`-Xprint:frontend`:
```scala
package <empty> {
  final lazy module val X: X$ = new X$()
  final module class X$() extends Object() { this: X.type => 
    @1 @inline() inline def foo: Int = 1
    def test: Unit = 
      {
        {
          X.foo
          ()
        }
      }
  }
}
```